Add share button for invocations

ahmedhamidawan opened the pull request:
Adds a button to the header for invocations which opens a modal where the user can click and share both the workflow and history associated with the invocation, and then also copy the link to the invocation to their clipboard.

Fixes https://github.com/galaxyproject/galaxy/issues/19756

https://github.com/user-attachments/assets/608b8bfb-f40a-4e19-8f13-311ea9b81b35

<details>
<summary><h3>Initial Progress-Wizard Implementation:</h3></summary>

This initially used the `GenericWizard` to enforce the user to make their history and workflow both accessible, and then showed them the link they can copy.

https://github.com/user-attachments/assets/4a006c8c-b04d-4d52-b9fa-428a42c73d8d

</details>

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

ahmedhamidawan committed:
Add sharing option for invocations

Opens a modal that does the following on `ok`:
- Share workflow (make importable)
- Share history (make importable)
- Copy the link to your clipboard

Fixes https://github.com/galaxyproject/galaxy/issues/19756

ahmedhamidawan commented on client/src/components/WorkflowInvocationState/WorkflowInvocationShare.vue:
```diff
@@ -0,0 +1,97 @@
+        return;
+    }
+
+    // Note: Is it worth it to check here if the workflow and history are already shareable?
```
🤔 ?

Also, now that I think of it, if both of those are already importable, should we even not open up a modal and instead just copy the link to clipboard and toast, and that's it?

mvdbeek replied:
yes, that would be good!

ahmedhamidawan replied:
✅ Done, and screencast updated in PR description.

Though, as mentioned in the commit message:
> **Note:** We will mostly have a summarized history object without the `importable` property and in that case we still attempt to make it shareable.
Only case I can see right now where the history _will_ have the `importable` property is when the history is current (and hence, we have a detailed object).

ahmedhamidawan committed:
parse error message better for workflow/history share

ahmedhamidawan committed:
add computed `owned` to `useWorkflowInstance`, and replace in components

ahmedhamidawan committed:
shorten button text for invocation sharing

ahmedhamidawan committed:
invocation share: only make history and workflow shareable if not already

Note: We will mostly have a summarized history object without the `importable` property and in that case we still attempt to make it shareable. Only case I can see right now where the history will have the `importable` property is when the history is current (and hence, we have a detailed object).

ahmedhamidawan committed:
do not even open modal if both workflow and history are shared

ahmedhamidawan committed:
add jest for `WorkflowInvocationShare`

davelopez commented on client/src/components/WorkflowInvocationState/WorkflowInvocationShare.vue:
```diff
@@ -0,0 +1,142 @@
+const historyAlreadyShareable = computed(() => {
+    // Note: We will mostly have a summarized history object without the importable property
+    //       and in that case we still attempt to make it shareable.
+    return !!history.value && hasImportable(history.value) && history.value.importable;
```
```suggestion
    return Boolean(history.value) && hasImportable(history.value) && history.value.importable;
```

Very minor comment, but I've seen this a couple of times and wanted to raise awareness about the cool style guide that we have :wink: 

https://github.com/galaxyproject/galaxy/blob/dev/client/docs/styleguide.md#casting

ahmedhamidawan replied:
That's a good point, thank you, I will make sure to do these right

ahmedhamidawan replied:
However, this causes this:
![image](https://github.com/user-attachments/assets/1355cf98-b787-44f3-87db-cc06fa894420)

Does this work better instead then?:
```suggestion
    return history.value && hasImportable(history.value) && history.value.importable;
```

davelopez replied:
Either that or you can improve `hasImportable` to handle null values too: `hasImportable(entry?: AnyHistory | null)`, whatever feels better to you :)

ahmedhamidawan replied:
https://github.com/galaxyproject/galaxy/pull/19959/commits/3534cae7e1b6b4ef19cb70bbe2beb08867a0c672

Please let me know if the changes here seem correct? 😃

davelopez replied:
Looks good! thank you!

ahmedhamidawan committed:
replace usage of `!!` with `Boolean` casting

Co-authored-by: David López <46503462+davelopez@users.noreply.github.com>

mvdbeek reviewed (approved):
Very cool, thank you @ahmedhamidawan!
